pydantic/decorators.py

```diff
@@ -131,67 +199,107 @@ def validator(
     ...
 
 
 @overload
 def root_validator(
     *,
-    mode: Literal['before', 'after', 'wrap', 'plain'] = 'after',
-    allow_reuse: bool = False,
-) -> Callable[[Callable[..., Any]], classmethod[Any]]:
+    # for pre=False you don't need to specify `skip_on_failure`
+    # if you specify `pre=True` then you don't need to specify
+    # `skip_on_failure`, in fact it is not allowed as an argument!
+    pre: Literal[True],
+    skip_on_failure: Literal[True],
+    allow_reuse: bool = ...,
+) -> Callable[[_V1RootValidatorFunctionType], _V1RootValidatorFunctionType,]:
     ...
 
 
+@overload
 def root_validator(
-    __func: Callable[..., Any] | None = None,
     *,
-    mode: Literal['before', 'after', 'wrap', 'plain'] = 'after',
+    # if you explicitly specify `pre=False` then you
+    # MUST specify `skip_on_failure=True`
+    pre: Literal[False],
+    skip_on_failure: Literal[True] = ...,
+    allow_reuse: bool = ...,
+) -> Callable[[_V1RootValidatorFunctionType], _V1RootValidatorFunctionType,]:
+    ...
+
+
+def root_validator(  # type: ignore[misc]
+    __func: _V1RootValidatorFunctionType | None = None,
+    *,
+    pre: bool = False,
+    skip_on_failure: bool = False,
     allow_reuse: bool = False,
-) -> classmethod[Any] | Callable[[Callable[..., Any]], classmethod[Any]]:
+) -> (
+    _decorators.PydanticDecoratorMarker[Any]
+    | Callable[
+        [_V1RootValidatorFunctionType],
+        _decorators.PydanticDecoratorMarker[Any],
+    ]
+):
     """
     Decorate methods on a model indicating that they should be used to validate (and perhaps modify) data either
     before or after standard model parsing/validation is performed.
     """
-    if __func:
-        f_cls = _decorators.prepare_validator_decorator(__func, allow_reuse)
-        setattr(f_cls, _decorators.ROOT_VALIDATOR_TAG, _decorators.Validator(mode=mode, is_field_validator=False))
-        return f_cls
-
-    def dec(f: Callable[..., Any]) -> classmethod[Any]:
-        f_cls = _decorators.prepare_validator_decorator(f, allow_reuse)
-        setattr(f_cls, _decorators.ROOT_VALIDATOR_TAG, _decorators.Validator(mode=mode, is_field_validator=False))
-        return f_cls
+    mode: Literal['before', 'after'] = 'before' if pre is True else 'after'
+    if pre is False and skip_on_failure is not True:
+        raise TypeError(
+            'If you use `@root_validator` with pre=False (the default)'
+            ' you MUST specify `skip_on_failure=True`.'
+            ' The `skip_on_failure=False` option is no longer available.'
+            ' If you were not trying to set `skip_on_failure=False` you'
+            ' can safely set `skip_on_failure=True`.'
+            ' If you do, this root validator will no longer be called'
+            ' if validation fails for any of the fields.'
+            ' Please see the migration guide for more details.'
+        )
+    wrap = partial(_decorators.make_v1_generic_root_validator, pre=pre)
+    if __func is not None:
```

Comment:
this branch is not covered in any tests

Reply:
thanks, I should have looked at coverage, removed now :)